### ISSUE TITLE
[BUG]: unscaled_grads don't work in sharded_optim_v2.py

### ISSUE THREAD

**Issue by VenAlone** [hpcaitech/ColossalAI#1218]
### 🐛 Describe the bug

Hi, I have great respect for this work! But when i debug the tests/test_zero/test_zero_engine.py, i found that on line 230 of sharded_optim_v2.py, the p.grad always equal None, scaling loss never works。And i also debug Colossalai-Examples/language/gpt/train_gpt.py,  unscale_grads function also don't work!

### Environment

cossalai==v0.1.7
torch==1.11.0+cu113

**Comment by VenAlone** [comment]
And i found that on line 285 in sharded_model_v2.py, all parameter's grad set to None，so unscaled_grad function will works abnormlly.

**Comment by VenAlone** [comment]
@[feifeibear](https://github.com/hpcaitech/ColossalAI/commits?author=feifeibear)
Maybe self._prepare_grads() should be called before unscaled_grads?Likes bellow:
`if self._prepare_grads_flag == False
    self._prepare_grads()
do unscaled grads
`

**Comment by ver217** [comment]
> @ver217 There is another bug!On line 185 of sharded_optim_v2.py, if using pipeline parallelism training or gradient accumulation，the backward will be called many times, but the self.optim_state flag set to OptimState.SCALED at first called, this means loss will be scaled only once time!

Why is the loss scaled only once? Each time you call `optimizer.backward(loss)`, it will be scaled.

**Comment by VenAlone** [comment]
> 

Because loss scaled only works with self.optim_state != OptimState.SCALED, once first calling optimizer.backward(loss), self.optim_state will set to OptimState.SCALED, at secord called, the self.optim_state == OptimState.SCALED, loss can't be scaled!

**Comment by ver217** [comment]
> > 
> 
> Because loss scaled only works with self.optim_state != OptimState.SCALED, once first calling optimizer.backward(loss), self.optim_state will set to OptimState.SCALED, at secord called, the self.optim_state == OptimState.SCALED, loss can't be scaled!

ShardedOptim does not have this limit. Each time you call `optimizer.backward(loss)`, it will be scaled.

**Comment by VenAlone** [comment]
> 

Another question, on line 172 in sharded_optim_v2.py, When using pipeline parallelism，the grads received from last stage has been scaled(last stage model has scaled loss, the received grad also be scaled)，backward_by_grads function don't set self.optim_state to OptimState.SCALED，so self.optim_state always equal to OptimState.UNSCALED if current stage isn't last stage，that means unscaled_grads will not works!

**Comment by ver217** [comment]
> > 
> 
> @ver217 Another question, on line 172 in sharded_optim_v2.py, When using pipeline parallelism，the grads received from last stage has been scaled(last stage model has scaled loss, the received grad also be scaled)，backward_by_grads function don't set self.optim_state to OptimState.SCALED，so self.optim_state always equal to OptimState.UNSCALED if current stage isn't last stage，that means unscaled_grads will not works!

Yes, you're right! Thanks for your reply. I've updated #1226
